doc/architecture/decisions/0008-add-first-class-citizen-fields-to-applications.md

```diff
@@ -0,0 +1,122 @@
+# 8. Add First-Class Citizen Fields to Applications
+
+Date: 2022-10-31
+
+## Status
+
+Accepted
+
+## Context
+
+When processing an application, there are some situations where we need easy
+access to particular fields within the JSON application object. These include
+(but are not limited to):
+
+- If an application is for a women-only AP
+- If an application is for a PIPE placement
+- The target location
+- The desired start date
+
+There are a number of options we could implement to do this. We could either:
+
+- save these additional fields to the database when an application is updated
+- update the UI code to send these additional fields to the API; or
+- use a [JsonLogic](https://jsonlogic.com) statement to fetch these items as part of the transformation process
+
+All of these approaches come with advantages and drawbacks:
+
+- saving the additional fields will require work at the API side, we would have to query the JSON object whenever it's updated to draw out the relevant fields. We would also have to consider what happens when the schema is updated
+- updating the UI code would further complicate the code at this level. We would have to update our page objects to pull out this data and marshal it into the correct format to send to the API, which would require extra conditionals and complexities. We also have no guarantee that the data would remain in sync.
+- using the JsonLogic statement in the transformer would require work to ensure this was updated every time the schema changed
+
+## Decision
+
+With all these approaches in mind, we have decided to use the transformers to
+add these additional fields to API responses using [JsonLogic](https://jsonlogic.com).
+
+Each schema entry in the database will also have a column in a JsonLogic
+format, mapping where to fetch the field from the JSON document. We can
+then process this query in the transformer. There are a couple of JsonLogic
+parsers we can use in a Kotlin application, such as [JSON Logic Java](https://github.com/jamsesso/json-logic-java),
+and [JSON Logic Kotlin](https://github.com/advantagefse/json-logic-kotlin).
+
+### Examples
+
+Assuming there is a `sentenceType` field in the JSON document
+that looks like this:
+
+```json
+{
+   "basic-information": {
+      "sentence-type": {
+         "sentenceType": "extendedDeterminate"
+      }
+   }
+}
+```
```

Comment:
Probably I'm not understanding properly as this feels like it might be coming across as more complex than it is?

Is it worth stepping through an example of through one of your listed "first class" data fields, such as `is_pipe_application` as another illustration?

The objective, as I understand it, is to keep two data values in lockstep, e.g:

### 1. `applications.is_pipe_application`: 
A simple property on the `applications` table which can readily be queried and indexed e.g. i) to select PIPE applications for allocation to appropriately qualified assessors, ii) to support indexing by Elasticsearch in preparation for matching against the trio of search dimensions: characteristics, availability and location, etc

### 2. `basic-information.ap-type` in JSON `data`
Deep in the JSON `data` blob , where it represents an answer to a particular questions for an in-progress application. 

### Requirements 
- Update [1] if [2] is changed. This can only be done before the application is submitted and whilst the schema version is unchanged
- Leave [1] unaffected if the schema version changes after the application has been submitted. 

So, does that boil down to needing a way of mapping a list of JSON paths to fields on the `applications` table. And that whenever the `data` blob is saved we want to update those mapped fields on `applications`? And to do that we'll need some code (such as JSONLogic) to map `ap-type == 'PIPE'` to a boolean on `applications.is_pipe_application`?

Reply:
The motivation here is to not have to do the first requirement. Ideally we don't want to have to update multiple fields with the same data, as I think that'll make things brittle. If we update the column type to [JSONB](https://www.compose.com/articles/faster-operations-with-the-jsonb-data-type-in-postgresql/), querying by any arbitary field will be easy and relatively fast, but we won't get the advantage of easy access to fields, but maybe that doesn't matter?